KEP-4622: Add docs for TopologyManager policy option for MaxAllowableNUMANodes (#46870)
kubernetes/website: content/en/docs/tasks/administer-cluster/topology-manager.md

Reviewer: Ideally, fix the formatting so that “_The following policy options exists:_” is followed by a _single_ unordered list.

You will still have to enable each option using the `TopologyManagerPolicyOptions` kubelet option.

The following policy options exists:

> * `prefer-closest-numa-nodes` (beta, and enabled by default).

Reviewer 2: I'd change to "beta, and visible by default".

The current unfortunately still not correct. In the context of the options, this means the feature is `visible` by default. Otherwise, you need to enable the feature gate to be able to set the option, and in turn the option default is the backward compatible one.

So it's a 2-level protection:
1. enable the option visibility (default: visible for beta, *not* visible for alpha)
2. enable the option proper

to summarize, the beta options are visible (not enabled) by default.

Sorry for any confusion caused.

Author: @ffromani this one I changed by accident and is part of another KEP.

  The `prefer-closest-numa-nodes` policy option is beta in Kubernetes {{< skew currentVersion >}}.

  If the `prefer-closest-numa-nodes` policy option is specified, the `best-effort` and `restricted`
  policies will favor sets of NUMA nodes with shorter distance between them when making admission decisions.
  You can enable this option by adding `prefer-closest-numa-nodes=true` to the Topology Manager policy options.
  By default, without this option, Topology Manager aligns resources on either a single NUMA node or
  the minimum number of NUMA nodes (in cases where more than one NUMA node is required). However,
  the `TopologyManager` is not aware of NUMA distances and does not take them into account when making admission decisions.
  This limitation surfaces in multi-socket, as well as single-socket multi NUMA systems,
  and can cause significant performance degradation in latency-critical execution and high-throughput applications if the
  Topology Manager decides to align resources on non-adjacent NUMA nodes.

> * `max-allowable-numa-nodes` (beta, and enabled by default).

Reviewer 2: likewise, `visible by default`

* `prefer-closest-numa-nodes` (beta, visible by default; `TopologyManagerPolicyOptions` and `TopologyManagerPolicyBetaOptions` feature gates have to be enabled).
  The `prefer-closest-numa-nodes` policy option is beta in Kubernetes {{< skew currentVersion >}}.

  If the `prefer-closest-numa-nodes` policy option is specified, the `best-effort` and `restricted`
  policies will favor sets of NUMA nodes with shorter distance between them when making admission decisions.
  You can enable this option by adding `prefer-closest-numa-nodes=true` to the Topology Manager policy options.
  By default, without this option, Topology Manager aligns resources on either a single NUMA node or
  the minimum number of NUMA nodes (in cases where more than one NUMA node is required). However,
  the `TopologyManager` is not aware of NUMA distances and does not take them into account when making admission decisions.
  This limitation surfaces in multi-socket, as well as single-socket multi NUMA systems,
  and can cause significant performance degradation in latency-critical execution and high-throughput applications if the
  Topology Manager decides to align resources on non-adjacent NUMA nodes.

> * `max-allowable-numa-nodes` (beta; `TopologyManagerPolicyOptions` and `TopologyManagerPolicyBetaOptions` feature gates have to be enabled)

Reviewer 2: ok, this is not related to this very change because I see the same text at line 225: this is not completely correct. The BetaOptions gate is always enabled by design. Please see https://github.com/kubernetes/kubernetes/blob/master/pkg/features/kube_features.go#L822 and https://github.com/kubernetes/kubernetes/blob/master/pkg/features/kube_features.go#L1200 .
IOW, the gates are currently enabled by default so no action is needed by default

>   The `max-allowable-numa-nodes` policy option is alpha in Kubernetes {{< skew currentVersion >}}.

Reviewer 3: `max-allowable-numa-nodes` is in beta stage.

Author: Done!

>   The value of `max-allowable-numa-nodes` does not (in and of itself) affect the latency of pod admission. With the `TopologyManager` enabled,

Reviewer 2: With the `TopologyManager` enabled -> When the Topology Manager policy is not `none`...
because the TopologyManager graduated to GA, so it's always enabled. However, setting the policy to `none` will bypass almost all its logic, rendering it a NOP

>   The value of `max-allowable-numa-nodes` does not (in and of itself) affect the latency of pod admission. With the `TopologyManager` enabled,

Reviewer 2: `The value of `max-allowable-numa-nodes` does not (in and of itself) affect the latency of pod admission.`
I'm not a native english speaker, but I'd put this detail at the end, once we described what the option do.

  the time to admit a pod is tied to the number of NUMA nodes on the physical machine. With the new `max-allowable-numa-nodes` option, nodes

>   with more than 8 NUMA nodes can be alowed to run with the `TopologyManager` enabled. However, it is unknown exactly how much this will slow

Reviewer 4: alowed -> allowed

Does this mean there is a soft threshold for `max-allowable-numa-nodes`?

Reviewer 2: TL;DR: we don't know yet.

the gist of this option is that the limit was added in the first place when machines with > 8 NUMA nodes were very rare, so the cap was set "high enough" because there wasn't enough data about how the topology manager algorithm scale.

Nowadays machines with > 8 NUMA nodes are common, and the kubelet just errors out in this scenario (https://github.com/kubernetes/kubernetes/issues/124144 and https://github.com/kubernetes/kubernetes/blob/release-1.30/pkg/kubelet/cm/topologymanager/topology_manager.go#L155 - this is in the setup stage)

So: we don't know how much the admission will slow down with more NUMA nodes to deal with. There's no enough data. Users are now enabled to at least try instead of getting a runtime error (and kubelet crash).

and can cause significant performance degradation in latency-critical execution and high-throughput applications if the
Topology Manager decides to align resources on non-adjacent NUMA nodes.

* `max-allowable-numa-nodes` (beta, visible by default).
The `max-allowable-numa-nodes` policy option is beta in Kubernetes {{< skew currentVersion >}}.

With the `TopologyManager` enabled, the time to admit a pod is tied to the number of NUMA nodes on the physical machine.

> With the new `max-allowable-numa-nodes` option, nodes with more than 8 NUMA nodes can be alowed to run

Reviewer 5: ```suggestion
With the new `max-allowable-numa-nodes` option, nodes with more than 8 NUMA nodes can be allowed to run
```

Reviewer: We also (strongly) avoid “new“; we like documentation to be [timeless](https://developers.google.com/style/timeless-documentation).

I'd make an exception for a page that was entirely about an alpha feature, because the quality bar is lower for alpha docs.

with the `TopologyManager` enabled. However, it is unknown exactly how much this will slow

> down pod admission on any given system. This feature is therefore to be used at-your-own-risk until we have

Reviewer 5: ```suggestion
down pod admission on any given system. This feature is therefore to be used at your own risk until we have
```

> With the `TopologyManager` enabled, the time to admit a pod is tied to the number of NUMA nodes on the physical machine.
> With the new `max-allowable-numa-nodes` option, nodes with more than 8 NUMA nodes can be alowed to run
> with the `TopologyManager` enabled. However, it is unknown exactly how much this will slow
> down pod admission on any given system. This feature is therefore to be used at-your-own-risk until we have
> a proper solution in place to reduce the state explosion that causes pod admission time to slow down as the
> number of NUMA nodes increases. The value of `max-allowable-numa-nodes` does not (in and of itself) affect the
> latency of pod admission.

Reviewer: ```suggestion
The time to admit a pod is tied to the number of NUMA nodes on the physical machine.
If you select the the `max-allowable-numa-nodes` policy option, nodes with more than 8 NUMA nodes can
be allowed to run with the topology enabled. The Kubernetes project only has limited data of the impact
of using the topology manager on (Kubernetes) nodes with more than 8 NUMA does. Because of that
lack of data, using this policy option is **not** recommended and is at your own risk.

Setting a value of `max-allowable-numa-nodes` does not (in and of itself) affect the
latency of pod admission, but binding a Pod to a (Kubernetes) node with many NUMA does does
have an impact.
Future, potential improvements to Kubernetes may improve Pod admission performance and the high
latency that happens as the number of NUMA nodes increases.
```

We usually [avoid statements about the future](https://kubernetes.io/docs/contribute/style/style-guide/#avoid-statements-about-the-future); this one is couched so that it aligns with the spirit of that guidance.
